umbraco/Umbraco-CMS
# Save media reference instead of media url

fixes https://github.com/umbraco/Umbraco-CMS/issues/4940

Links to media is now stored as localLink in  the RTE and Grid RTE and the URL is resolved when rendering, img tags are already handled (see https://github.com/umbraco/Umbraco-CMS/commit/99b80e0f569ce24623eb331fd683b5255a64c672)

The migration looks for all links that starts with `/media/`  and if it can be looked up in the DB the value is replaced with the UDI

### Breaking changes:

RTE and RTE in Grid now stores localLink for media like it does for content.

Before:
```html
<p><a data-udi="umb://media/662af6ca411a4c93a6c722c4845698e7" href="/media/662af6ca411a4c93a6c722c4845698e7/00000006000000000000000000000000/16403439029_f500be349b_o.jpg" title="Umbraco Campari Meeting Room">Umbraco Campari Meeting Room</a></p>
```
After:
```html
<p><a data-udi="umb://media/662af6ca411a4c93a6c722c4845698e7" href="/{localLink:umb://media/662af6ca411a4c93a6c722c4845698e7}" title="Umbraco Campari Meeting Room">Umbraco Campari Meeting Room</a></p>
```
The url's are resolved when rendering using TemplateUtilities.ParseInternalLinks which the RTE value converter calls and it's used in the default Grid/Editors/Rte.cshtml file.

### How to test:

* Add links to media in both RTE and Grid RTE check the source and check that the link is a `localLink` e.g. `/{loclLink:umb://media/id}`
* Render the properties on a page and check the links are resolved to the correct media URL
* To test the migration add the links before checking out this branch and check the source and ensure that the links are referencing the `/media/` path, the checkout the branch and let the migration run then check the source and verify the links has been replaced by a `localLink`
